nextcloud/server
# Add scheduling availability settings

Fixes https://github.com/nextcloud/server/issues/27422

![Bildschirmfoto von 2021-10-18 20-17-39](https://user-images.githubusercontent.com/1374172/137785462-a6451137-6c0e-4b73-8673-a795e6e493e1.png)
![Bildschirmfoto von 2021-10-18 20-18-52](https://user-images.githubusercontent.com/1374172/137785472-404ecf75-a122-452c-9d4f-d7a5bc2d0db5.png)


<details>

This is an early prototype for https://github.com/nextcloud/server/issues/27422

![Bildschirmfoto von 2021-06-10 20-08-36](https://user-images.githubusercontent.com/1374172/121576419-a58e4880-ca28-11eb-9711-35c43ba522b3.png)

</details>

To do
- [x] Fetch current availability value form the scheduling inbox via a propfind
- [x] Ability to select a timezone
  * Auto-detect default via jstz
- [x] Ability to add another slot for a day
- [x] Ability to remove a slot from a day, but not the last one
- [x] Prop patch the scheduling inbox with the updated availability rrule
- [x] Start the week at Mo/Su depending on locale

Steps to test
* Open the new settings
* Set some slots where you want to be available and save them
* Open the Calendar app
* Create an event, invite another user or email address and open the free/busy view
* You should see anything outside the availability slots as *out of office* (screenshot above)

Ref https://datatracker.ietf.org/doc/html/rfc7953

Follow-ups
* Deduplicate slots into blocked AVAILABLE components if possible, e.g. when 9-5 is used five days a week
* https://github.com/nextcloud/server/pull/29318 Save and restore selected timezone as a component in the VAVAILABILITY
* https://github.com/nextcloud/server/issues/29319 Possible design polishing
* Error handling of unprocessible VAVAILABILITY values set by other applications
